instrumentation/opentelemetry-instrumentation-botocore/src/opentelemetry/instrumentation/botocore/__init__.py

```diff
@@ -44,6 +44,40 @@
 
 API
 ---
+
+The `instrument` method accepts the following keyword args:
+
+tracer_provider (TracerProvider) - an optional tracer provider
+request_hooks (dict) - a mapping between service names their respective callable request hooks
+* a request hook signature is: def request_hook(span: Span, operation_name: str, api_params: dict) -> None
+response_hooks (dict) - a mapping between service names their respective callable response hooks
+* a response hook signature is: def response_hook(span: Span, operation_name: str, result: dict) -> None
+
+for example:
+
+.. code: python
+
+    from opentelemetry.instrumentation.botocore import BotocoreInstrumentor
+    import botocore
+
+    def ec2_request_hook(span, operation_name, api_params):
+        # request hook logic
+
+    def ec2_response_hook(span, operation_name, result):
+        # response hook logic
+
+    # Instrument Botocore with hooks
+    BotocoreInstrumentor().instrument(
+        request_hooks={"ec2": ec2_request_hook}, response_hooks={"ec2": ec2_response_hook}
```

Comment:
This is deviating a bit too much from other instrumentation packages. How important is to map hooks to sub-components? Is there a very compelling use case for this?

What if we just accepting a single request/response hook and passed the "service name" as an argument to it? Users would then have some logic branching based on the service name. I'd imagine it wouldn't make it too terrible to implement hooks that handle different services in different ways but it'd simplify it for the common case where people just want a simple, single hook for a given instrumentation package. What do you think?

Reply:
I think you are right

My idea here was to let the users pass hooks, without the need to write the logic of branching based on the service name, but since they anyway have to create a dictionary with a mapping between service names to the respective code, it won't require much more work to write this logic branching.

I'll change it to a single request hook and single response hook, to make it consistent with the other libraries